### ISSUE TITLE
Add option to make config value more stricitive

### ISSUE THREAD

**Issue by ganmacs** [fluent/fluentd#2616]
**Is your feature request related to a problem? Please describe.**

ref: https://github.com/fluent/fluentd/issues/2607

**Describe the solution you'd like**

https://github.com/fluent/fluentd/pull/2615#issuecomment-530246816

> We should improve this point to use Integer() or others.
This change has compatibility probolem so it will be implemented in v2 or via configuration.

**Additional context**

**Comment by ashie** [comment]
In #2685, I've implemented `integer`, `float` and `bool` cases.

> In addition, support nil is better.
> I sometimes see users use "#{ENV['SOME_ENV_VAR']}" in the configuration and they expect nil is  passed when no envvar. Currently, this is converted into "".

I'm now considering this case.
IMHO it shouldn't be treated as `nil`, empty string `""` is the correct behavior.
I think their expectation is wrong because [the document](https://docs.fluentd.org/configuration/config-file#embedding-ruby-expressions) says

```
You can evaluate the Ruby code with #{} in " quoted string.
```

and `#{ENV['SOME_ENV_VAR']}` is evaluated as empty string in actual Ruby code.

```
irb(main):001:0> "#{ENV['SOME_ENV_VAR']}"
=> ""
```

so I don't expect `nil` for it.
If the literal is `ENV['SOME_ENV_VAR']` (without `#{}`), I'll expect `nil` for it though:

```
irb(main):001:0> ENV['SOME_ENV_VAR']
=> nil
```

**Comment by ashie** [comment]
Is there any other case that `nil` should be supported?

**Comment by ganmacs** [comment]
> In addition, support nil is better.

What does this expect in this feature? 
`nil` can pass the validation even if the value is expected to be something type(String, Integer or whatever)?

**Comment by cosmo0920** [comment]
> Is there any other case that `nil` should be supported?

Handling for `ENV['NONEXISTENT']` case?

**Comment by repeatedly** [comment]
> I think their expectation is wrong because the document says

Yes but many users think it becomes `nil` and this is why users use `username "#{ENV['XXX_USER'] || nil}"` in many places. Of course, this doesn't work as expected.

Other apporch is adding `<if ruby cond>` approch to support ENV usecase like below.

```
<if ENV['XXX_USER']>
  username "#{ENV['XXX_USER']}"
</if>
```

**Comment by ashie** [comment]
Thanks for your comment.

> Other apporch is adding <if ruby cond> approch to support ENV usecase like below.
> 
> ```
> <if ENV['XXX_USER']>
>   username "#{ENV['XXX_USER']}"
> </if>
> ```

You mean that the parameter should be treated as "unspecified" (use the default value) if the value is 'nil', don't you? It's an interesting feature.

I think it should be treated as a different issue, it's a different feature from #2685.
We should determine the syntax for `nil`.
I'll file a new issue after completing #2685.

**Comment by ganmacs** [comment]
> I think it should be treated as a different issue, it's a different feature from #2685.

+1

@repeatedly Could you tell me your thought?  I don't still understand what is "support nil".
https://github.com/fluent/fluentd/issues/2616#issuecomment-553729392

**Comment by ashie** [comment]
BTW

```
this is why users use username "#{ENV['XXX_USER'] || nil}" in many places. Of course, this doesn't work as expected.
```

Following dirty hacks may be effective for such case...

```
   username "#{ENV['XXX_USER'] || raise("no username")}"
```

```
   # launcher can detect it by the exit status
   username "#{ENV['XXX_USER'] || exit(123)}"
```

**Comment by ganmacs** [comment]
Ruby has Hash#fetch(https://apidock.com/ruby/Hash/fetch)
Isn't `username "#{ENV.fetch('XX_USER')}"` enough for such case?

**Comment by repeatedly** [comment]
Above works? I tested it and fluentd stopped by ```/path/to/lib/fluent/config/literal_parser.rb:171:in `instance_eval': key not found: "NUM" (KeyError)``` exception.

> You mean that the parameter should be treated as "unspecified" (use the default value) if the value is 'nil', don't you? It's an interesting feature.

No. I show `nil` because almost plugins use `nil` for default value like below.

```
config_param :str_param, :string, :default => nil
config_param :num_param, :integer, :default => nil
```

Currently, `str "#{ENV['STR'] || nil}"` becomes `''` and `num "#{ENV['NUM'] || nil}"` becomes `0` in plugin code but users think this becomes `nil`. We need to fix this gap.

I just notice helper method is also one adhoc approach like below:

```
 `num "#{ENV['NUM'] || use_default}"` or ``num "#{ENV['NUM'] || use_nil}"``
```

**Comment by ashie** [comment]
```
num_param "#{ENV['NUM'] || use_default}"  # raise specific exception to skip value processing. use default in config_param.
or
num_param "#{ENV['NUM'] || use_nil}"
```

I'm now confirming this concept in another branch:
https://github.com/ashie/fluentd/commit/f7b5914f5fa3a1fab240ee61508b0bf824d97b85

Implementing 'use_nil' seems possible, but 'use_default' seems difficult because default value isn't determined while parsing the config file. It's determined at Configurable::configure.

**Comment by ashie** [comment]
> but 'use_default' seems difficult because default value isn't determined while parsing the config file. It's determined at Configurable::configure which is run after parsing config file (receives a parsed config as an argument).

Now I've noticed that it's possible by returning ":default" or something from `LiteralParser` and processing it at `Config::STRING_TYPE`.

**Comment by ashie** [comment]
> I'm now confirming this concept in another branch:
> https://github.com/ashie/fluentd/commit/f7b5914f5fa3a1fab240ee61508b0bf824d97b85

I've merged these changes to #2685, since they are based on changes in #2685.
